---
title: periodicDNA
author: Jacques Serizay
date: "`r Sys.Date()`"
output: rmarkdown::html_vignette
vignette: >
    %\VignetteIndexEntry{Introduction to periodicDNA}
    %\VignetteEngine{knitr::rmarkdown}
    %\VignetteEncoding{UTF-8}
---

```{r, eval = TRUE, echo=FALSE, results="hide", warning=FALSE}
knitr::opts_chunk$set(
    collapse = TRUE,
    comment = "#>", 
    fig.path = "man/figures/periodicDNA-"
)
suppressPackageStartupMessages({
    library(GenomicRanges)
    library(magrittr)
    library(ggplot2)
    library(periodicDNA)
})
```

### Dinucleotide periodicity over a set of Genomic Ranges

The main goal of periodicDNA is to quantify periodicity of a given k-mer in a
set of sequences. `getPeriodicity()` can be used to achieve this. 
In the following example, `getPeriodicity()` is usually ran using 
a GRanges object, specifying from which genome this GRanges comes from. 

```{r, eval = TRUE}
data(proms)
periodicity_result <- getPeriodicity(
    proms,
    genome = 'ce11',
    motif = 'TT', 
    cores = 1
)
```

The output of `getPeriodicity()` is an object containing the PSD scores 
associated with each period. For a given period, a high PSD score 
indicates a high periodicity of the k-mer of interest. 

```{r, eval = TRUE}
head(periodicity_result$PSD)
```

Graphical output of `getPeriodicity()` can be obtained using the 
`plotPeriodicityResults()` function: 

```{r, eval = FALSE}
plotPeriodicityResults(periodicity_result)
```

![](https://raw.githubusercontent.com/js2264/periodicDNA/master/man/figures/ubiquitous-promoters_TT-periodicity.png)

The first plot shows the raw distribution of distances between pairs of 'TT' 
in the sequences of the provided GRanges. The second plot shows the 
decay-normalised distribution. Finally, the third plot shows the PSD scores 
of the 'TT' k-mer, measured from the normalised distribution.  

**Note:**  
`getPeriodicity()` can also be ran directly on a set of sequences of interest 
as follows: 

```{r, eval = FALSE}
data(ce_proms_seqs)
periodicity_result <- getPeriodicity(
    ce_proms_seqs,
    motif = 'TT', 
    cores = 4
)
```

### Track of periodicity over a set of Genomic Ranges

Another major aim of periodicDNA is to generate continuous linear tracks of 
k-mer periodicity strength over genomic loci of interest. 
`getPeriodicityTrack()` can be used to generate suck genomic tracks. In the 
following example,  

```{r, eval = FALSE}
WW_10bp_track <- getPeriodicityTrack(
    Biostrings::getSeq(
        BSgenome.Celegans.UCSC.ce11::BSgenome.Celegans.UCSC.ce11
    ),
    granges = proms, 
    motif = 'WW',
    period = 10,
    cores = 12, 
    bw_file = 'WW-10-bp-periodicity_over-proms.bw'
)
```

When plotted over sets of ubiquitous, germline or somatic TSSs, the resulting 
track clearly shows increase of WW 10-bp periodicity above the ubiquitous and
germline TSSs, whereas somatic TSSs do not show such increase.

```{r, eval = FALSE, results="hide", warning=FALSE}
data(ce11_TSSs)
plotAggregateCoverage(
    WW_10bp_track, 
    ce11_TSSs, 
    xlab = 'TSS',
    ylab = 'WW 10-bp PSD score'
)
```

![TT-periodicity](https://raw.githubusercontent.com/js2264/periodicDNA/master/man/figures/TT-10bp-periodicity_tissue-spe-TSSs.png)

### Fold Power Increase

Fold Power Increase (FPI) is a metric introduced by Pich et al., 
Cell 2018. It can be computed using the `getFPI()` function. In the 
following example, the FPI of TT dinucleotides at a 10-bp periodicity is 
computed over 200-bp long sequences randomly sampled in ce11 genome. 

```{r, eval = TRUE}
ce11_random_regions <- sampleGenome(
    'ce11',
    n = 200, w = 1000
)
fpi <- getFPI(
    ce11_random_regions$seq, 
    motif = 'TT', 
    cores_shuffling = 1, 
    n_shuffling = 10
)
plotFPI(fpi)
```

## Session info

```{r, eval = TRUE}
sessionInfo()
```