<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>periodicDNA • periodicDNA</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png">
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/sandstone/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="periodicDNA">
<meta property="og:description" content="periodicDNA">
<meta property="og:image" content="/logo.png">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">periodicDNA</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.3.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/periodicDNA.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/js2264/periodicDNA/">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>periodicDNA</h1>
                        <h4 class="author">Jacques Serizay</h4>
            
            <h4 class="date">2020-05-05</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/js2264/periodicDNA/blob/master/vignettes/periodicDNA.Rmd"><code>vignettes/periodicDNA.Rmd</code></a></small>
      <div class="hidden name"><code>periodicDNA.Rmd</code></div>

    </div>

    
    
<div id="introduction" class="section level2">
<h2 class="hasAnchor">
<a href="#introduction" class="anchor"></a>Introduction</h2>
<p>Many oligonucleotides are periodically occurring in genomes across eukaryotes, and some are impacting the physical properties of DNA. For example, DNA bendability is modulated by 10-bp periodic occurrences of WW (W = A/T) dinucleotides. periodicDNA is an R package offering a set of functions to identify k-mer periodicity in sequences of interest such as regulatory elements. It also provides a method to generate continuous tracks of k-mer periodicity over genomic loci.</p>
</div>
<div id="installation" class="section level2">
<h2 class="hasAnchor">
<a href="#installation" class="anchor"></a>Installation</h2>
<p>periodicDNA can be installed from GitHub as follows:</p>
<div class="sourceCode" id="cb1"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html">install.packages</a></span>(<span class="st">"devtools"</span>)
<span class="kw pkg">devtools</span><span class="kw ns">::</span><span class="fu"><a href="https://rdrr.io/pkg/devtools/man/reexports.html">install_github</a></span>(<span class="st">"js2264/periodicDNA"</span>)
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">periodicDNA</span>)</pre></body></html></div>
</div>
<div id="loading-periodicdna-and-other-useful-packages" class="section level2">
<h2 class="hasAnchor">
<a href="#loading-periodicdna-and-other-useful-packages" class="anchor"></a>Loading periodicDNA and other useful packages</h2>
<div class="sourceCode" id="cb2"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">periodicDNA</span>)
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">magrittr</span>)
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">ggplot2</span>)</pre></body></html></div>
</div>
<div id="how-periodicdna-works" class="section level2">
<h2 class="hasAnchor">
<a href="#how-periodicdna-works" class="anchor"></a>How periodicDNA works</h2>
<p>When running <code><a href="../reference/getPeriodicity.html">getPeriodicity()</a></code> function for a given k-mer (typically a dinucleotide) over a set of GRanges, R does the following steps:</p>
<ol style="list-style-type: decimal">
<li>It first finds all the possible motif occurrences over the GRanges;<br>
</li>
<li>It then find all the possible pairwise distances and generates a normalised histogram of the pairwise distances;<br>
</li>
<li>It performs a Fast Fourier Transform of the normalised histogram;</li>
<li>Finally, it outputs the power spectrum density of the k-mer in the provided GRanges.</li>
</ol>
</div>
<div id="quick-k-mer-periodicity" class="section level2">
<h2 class="hasAnchor">
<a href="#quick-k-mer-periodicity" class="anchor"></a>Quick k-mer periodicity</h2>
<p>A genome sequence and a set of genomic loci must be defined. Let’s focus on the C. elegans genome for now, and more specifically around its ubiquitous TSSs.</p>
<div class="sourceCode" id="cb3"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/utils/data.html">data</a></span>(<span class="no">ce_proms</span>)
<span class="no">ubiq_TSSs</span> <span class="kw">&lt;-</span> <span class="no">ce_proms</span>[<span class="no">ce_proms</span>$<span class="no">which.tissues</span> <span class="kw">==</span> <span class="st">'Ubiq.'</span>] <span class="kw">%&gt;%</span>
    <span class="fu"><a href="../reference/deconvolveBidirectionalPromoters.html">deconvolveBidirectionalPromoters</a></span>() <span class="kw">%&gt;%</span>
    <span class="fu"><a href="../reference/alignToTSS.html">alignToTSS</a></span>(<span class="fl">50</span>, <span class="fl">300</span>)
<span class="no">seqs</span> <span class="kw">&lt;-</span> <span class="fu">getSeq</span>(
    <span class="kw pkg">BSgenome.Celegans.UCSC.ce11</span><span class="kw ns">::</span><span class="no"><a href="https://rdrr.io/pkg/BSgenome.Celegans.UCSC.ce11/man/package.html">BSgenome.Celegans.UCSC.ce11</a></span>
)
<span class="no">ubiq_seqs</span> <span class="kw">&lt;-</span> <span class="no">seqs</span>[<span class="no">ubiq_TSSs</span>]
<span class="no">ubiq_TT</span> <span class="kw">&lt;-</span> <span class="fu"><a href="../reference/getPeriodicity.html">getPeriodicity</a></span>(
    <span class="no">ubiq_seqs</span>,
    <span class="kw">motif</span> <span class="kw">=</span> <span class="st">'TT'</span>,
    <span class="kw">cores</span> <span class="kw">=</span> <span class="fl">1</span>
)
<span class="co">#&gt; - Mapping k-mers.</span>
<span class="co">#&gt; - 3668916 k-mers found.</span>
<span class="co">#&gt; - Calculating pairwise distances.</span>
<span class="co">#&gt; - Normalizing histogram vector.</span>
<span class="co">#&gt; - Applying Fast Fourier Transform to the vector of distances.</span>
<span class="co">#&gt; - Computing signal-to-noise ratios.</span>
<span class="fu"><a href="../reference/plotPeriodicityResults.html">plotPeriodicityResults</a></span>(<span class="no">ubiq_TT</span>)
<span class="co">#&gt; [[1]]</span></pre></body></html></div>
<p><img src="periodicDNA_files/figure-html/unnamed-chunk-4-1.png" width="700"></p>
<pre><code>#&gt; 
#&gt; [[2]]</code></pre>
<p><img src="periodicDNA_files/figure-html/unnamed-chunk-4-2.png" width="700"></p>
<pre><code>#&gt; 
#&gt; [[3]]</code></pre>
<p><img src="periodicDNA_files/figure-html/unnamed-chunk-4-3.png" width="700"></p>
<p><code>getPeriodicity</code> can also works with GRanges objects, as long as a <code>genome</code> is supplied. ‘ce11’, ‘sacCer3’, ‘mm10’, ‘dm6’, ‘hg38’ and ‘danRer10’ are currently supported. Otherwise, a BSgenome can be used directly.</p>
<div class="sourceCode" id="cb6"><html><body><pre class="r"><span class="no">ubiq_TT</span> <span class="kw">&lt;-</span> <span class="fu"><a href="../reference/getPeriodicity.html">getPeriodicity</a></span>(
    <span class="no">ce_proms</span>,
    <span class="kw">genome</span> <span class="kw">=</span> <span class="st">'ce11'</span>,
    <span class="kw">motif</span> <span class="kw">=</span> <span class="st">'TT'</span>,
    <span class="kw">cores</span> <span class="kw">=</span> <span class="fl">1</span>
)
<span class="co">#&gt; - Mapping k-mers.</span>
<span class="co">#&gt; - 2638289 k-mers found.</span>
<span class="co">#&gt; - Calculating pairwise distances.</span>
<span class="co">#&gt; - Normalizing histogram vector.</span>
<span class="co">#&gt; - Applying Fast Fourier Transform to the vector of distances.</span>
<span class="co">#&gt; Range (1:200) is wider than any range in the current </span>
<span class="co">#&gt;             distances vector. Shortening the range from 1 to 149...</span>
<span class="co">#&gt; - Computing signal-to-noise ratios.</span></pre></body></html></div>
<!--  
## Fold Power Increase metric

The FPI (fold power increase) metric was originally described in 
Pich et al., Cell 2018. It allows to accurately estimate the periodicity 
strength of a k-mer of interest by modelling the background periodicity.


```r
fpi <- getFPI(
    ubiq_seqs,
    motif = 'TT', 
    parallel_shuffling = 1, 
    n_shuffle = 10
)
#> - Calculating observed PSD
#> - Mapping k-mers.
#> - 3668916 k-mers found.
#> - Calculating pairwise distances.
#> - Normalizing histogram vector.
#> - Applying Fast Fourier Transform to the vector of distances.
#> - Computing signal-to-noise ratios.
#> >> Measured PSD @ 10bp is: 3.06609400342861e-06
#> - Shuffling 1/10
#> - Shuffling 2/10
#> - Shuffling 3/10
#> - Shuffling 4/10
#> - Shuffling 5/10
#> - Shuffling 6/10
#> - Shuffling 7/10
#> - Shuffling 8/10
#> - Shuffling 9/10
#> - Shuffling 10/10
plotFPI(fpi)
```

<img src="periodicDNA_files/figure-html/unnamed-chunk-6-1.png" width="700" />

**Note:** To calculate the FPI of a k-mer in a set of sequences, the sequences
of interest are first randomly shuffled (`n_shuffle` times). This process is
stochastic and for this reason, the FPI score might vary each time the 
`getFPI()` function is ran. To ensure accurate estimation of the FPI score, 
it is recommended to run the `getFPI()` function with a high `n_shuffle`. The 
process can be parallelised using the `parallel_shuffling` argument. 

-->
</div>
<div id="generating-periodicity-tracks-with-periodicdna" class="section level2">
<h2 class="hasAnchor">
<a href="#generating-periodicity-tracks-with-periodicdna" class="anchor"></a>Generating periodicity tracks with periodicDNA</h2>
<p>Rather than computing the periodicity of a chosen k-mer in a given set of sequences, <code><a href="../reference/generatePeriodicityTrack.html">generatePeriodicityTrack()</a></code> function computes a linear periodicity track covering loci of interest.</p>
<div class="sourceCode" id="cb7"><html><body><pre class="r"><span class="fu"><a href="../reference/generatePeriodicityTrack.html">generatePeriodicityTrack</a></span>(
    <span class="no">seqs</span>,
    <span class="kw">granges</span> <span class="kw">=</span> <span class="no">ce_proms</span>[<span class="fl">1</span>],
    <span class="kw">motif</span> <span class="kw">=</span> <span class="st">'TT'</span>,
    <span class="kw">freq</span> <span class="kw">=</span> <span class="fl">1</span>/<span class="fl">10</span>,
    <span class="kw">cores</span> <span class="kw">=</span> <span class="fl">1</span>
)
<span class="co">#&gt; </span>
<span class="co">#&gt;     The genome has been divided in 100-bp bins</span>
<span class="co">#&gt;     with a sliding window of 2-bp.</span>
<span class="co">#&gt;     </span>
<span class="co">#&gt;     549 windows overlap the input </span>
<span class="co">#&gt;     loci (extended to 1000-bp).</span>
<span class="co">#&gt;     </span>
<span class="co">#&gt;     The mapping will be split among 1 jobs.</span>
<span class="co">#&gt;     </span>
<span class="co">#&gt;     Each job will analyse 549 bins.</span>
<span class="co">#&gt;     </span>
<span class="co">#&gt;     Now starting...</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; &gt;&gt; 1 bins computed /// TIME: 2020-05-05 18:04:17</span>
<span class="co">#&gt; &gt;&gt; 549 bins computed /// TIME: 2020-05-05 18:04:39</span>
<span class="co">#&gt; </span>
<span class="co">#&gt;    SUCCESS: TT-periodicity_g-100^2_b-100^5.bw has been created!</span>
<span class="co">#&gt;    1098 bases covered by the generated track.</span>
<span class="co">#&gt; [1] TRUE</span></pre></body></html></div>
</div>
<div id="session-info" class="section level2">
<h2 class="hasAnchor">
<a href="#session-info" class="anchor"></a>Session info</h2>
<div class="sourceCode" id="cb8"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html">sessionInfo</a></span>()
<span class="co">#&gt; R version 3.5.2 (2018-12-20)</span>
<span class="co">#&gt; Platform: x86_64-pc-linux-gnu (64-bit)</span>
<span class="co">#&gt; Running under: Ubuntu 18.04.2 LTS</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Matrix products: default</span>
<span class="co">#&gt; BLAS: /usr/lib/x86_64-linux-gnu/blas/libblas.so.3.7.1</span>
<span class="co">#&gt; LAPACK: /usr/lib/x86_64-linux-gnu/lapack/liblapack.so.3.7.1</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; locale:</span>
<span class="co">#&gt;  [1] LC_CTYPE=en_GB.UTF-8       LC_NUMERIC=C              </span>
<span class="co">#&gt;  [3] LC_TIME=en_GB.UTF-8        LC_COLLATE=en_GB.UTF-8    </span>
<span class="co">#&gt;  [5] LC_MONETARY=en_GB.UTF-8    LC_MESSAGES=en_GB.UTF-8   </span>
<span class="co">#&gt;  [7] LC_PAPER=en_GB.UTF-8       LC_NAME=C                 </span>
<span class="co">#&gt;  [9] LC_ADDRESS=C               LC_TELEPHONE=C            </span>
<span class="co">#&gt; [11] LC_MEASUREMENT=en_GB.UTF-8 LC_IDENTIFICATION=C       </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; attached base packages:</span>
<span class="co">#&gt; [1] parallel  stats4    stats     graphics  grDevices utils     datasets </span>
<span class="co">#&gt; [8] methods   base     </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; other attached packages:</span>
<span class="co">#&gt;  [1] periodicDNA_0.3.0    Biostrings_2.50.2    XVector_0.22.0      </span>
<span class="co">#&gt;  [4] ggplot2_3.1.1        magrittr_1.5         GenomicRanges_1.34.0</span>
<span class="co">#&gt;  [7] GenomeInfoDb_1.18.2  IRanges_2.16.0       S4Vectors_0.20.1    </span>
<span class="co">#&gt; [10] BiocGenerics_0.28.0 </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; loaded via a namespace (and not attached):</span>
<span class="co">#&gt;  [1] SummarizedExperiment_1.12.0       zoo_1.8-5                        </span>
<span class="co">#&gt;  [3] tidyselect_0.2.5                  xfun_0.5                         </span>
<span class="co">#&gt;  [5] purrr_0.3.2                       lattice_0.20-38                  </span>
<span class="co">#&gt;  [7] colorspace_1.4-1                  htmltools_0.3.6                  </span>
<span class="co">#&gt;  [9] rtracklayer_1.42.2                yaml_2.2.0                       </span>
<span class="co">#&gt; [11] XML_3.98-1.19                     rlang_0.4.2                      </span>
<span class="co">#&gt; [13] pkgdown_1.5.1                     pillar_1.3.1                     </span>
<span class="co">#&gt; [15] glue_1.3.1                        withr_2.1.2                      </span>
<span class="co">#&gt; [17] BiocParallel_1.16.6               matrixStats_0.54.0               </span>
<span class="co">#&gt; [19] GenomeInfoDbData_1.2.0            plyr_1.8.4                       </span>
<span class="co">#&gt; [21] stringr_1.4.0                     zlibbioc_1.28.0                  </span>
<span class="co">#&gt; [23] munsell_0.5.0                     gtable_0.3.0                     </span>
<span class="co">#&gt; [25] memoise_1.1.0                     evaluate_0.13                    </span>
<span class="co">#&gt; [27] labeling_0.3                      Biobase_2.42.0                   </span>
<span class="co">#&gt; [29] knitr_1.22                        Rcpp_1.0.1                       </span>
<span class="co">#&gt; [31] BSgenome_1.50.0                   scales_1.0.0                     </span>
<span class="co">#&gt; [33] backports_1.1.3                   DelayedArray_0.8.0               </span>
<span class="co">#&gt; [35] desc_1.2.0                        fs_1.3.1                         </span>
<span class="co">#&gt; [37] Rsamtools_1.34.1                  digest_0.6.18                    </span>
<span class="co">#&gt; [39] stringi_1.3.1                     dplyr_0.8.0.1                    </span>
<span class="co">#&gt; [41] grid_3.5.2                        rprojroot_1.3-2                  </span>
<span class="co">#&gt; [43] tools_3.5.2                       bitops_1.0-6                     </span>
<span class="co">#&gt; [45] lazyeval_0.2.2                    RCurl_1.95-4.12                  </span>
<span class="co">#&gt; [47] tibble_2.1.1                      crayon_1.3.4                     </span>
<span class="co">#&gt; [49] pkgconfig_2.0.2                   BSgenome.Celegans.UCSC.ce11_1.4.2</span>
<span class="co">#&gt; [51] Matrix_1.2-17                     MASS_7.3-51.1                    </span>
<span class="co">#&gt; [53] assertthat_0.2.1                  rmarkdown_1.12.6                 </span>
<span class="co">#&gt; [55] R6_2.4.0                          GenomicAlignments_1.18.1         </span>
<span class="co">#&gt; [57] compiler_3.5.2</span></pre></body></html></div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by <a href="https://js2264.github.io/">Jacques Serizay</a>.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.5.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
